game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/AaDefenseCombatValue.java

```diff
@@ -0,0 +1,95 @@
+package games.strategy.triplea.delegate.power.calculator;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.attachments.UnitSupportAttachment;
+import java.util.function.Predicate;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Value;
+
+/**
+ * Calculates defense strength and roll for AA/Targeted dice
+ *
+ * <p>This takes into account friendly support, and enemy support
+ */
+@Builder
+@Value
+@Getter(AccessLevel.NONE)
+class AaDefenseCombatValue implements CombatValue {
+
+  @NonNull GameData data;
+  @NonNull AvailableSupports friendlySupportTracker;
+  @NonNull AvailableSupports enemySupportTracker;
+
+  @Override
+  public StrengthOrRollCalculator getRoll() {
+    return new AaDefenseRoll(friendlySupportTracker, enemySupportTracker);
+  }
+
+  @Override
+  public StrengthOrRollCalculator getStrength() {
+    return new AaDefenseStrength(data, friendlySupportTracker, enemySupportTracker);
+  }
+
+  @Override
+  public boolean isDefending() {
+    return true;
```

Comment:
It's a bit odd this class is not generic for either offense or defnese. Could you double check if it's possible to make the defense flag, a flag within the implementation.

Reply:
The name of this class is Aa**Defense**CombatValue.  It can't be generic since it is for defense.  So I'm not sure what I can do.

Reply:
If internalizing the flag, then the class would presumably would be merged with the offense combat value class. I somewhat like that as it keeps the AA combat value concerns together, seemingly they are very similar classes.

Reply:
This flag is/will be used to generate the `TotalPowerAndTotalStrength` objects.  I might be able to flip the construction around so that `TotalPowerAndTotalStrength` has the defense flag instead of needing to grab it from `CombatValue` but that will have to be a future change.

Reply:
And `AaOffenseStrength` and `AaDefenseStrength` are different.  I'll look at creating a parent class that shares the common roll calculator and just separates the strength calculators.

Reply:
I'm not sure a parent class will be the way to go. Have you considered injecting a strategy object that instead contains the common functionality?

- https://codeburst.io/inheritance-is-evil-stop-using-it-6c4f1caf5117

- https://softwareengineering.stackexchange.com/questions/260343/why-is-inheritance-generally-viewed-as-a-bad-thing-by-oop-proponents